pkg/jobs/hypershift/hypershift.go

```diff
@@ -0,0 +1,533 @@
+	}
+
+	return true
+}
+
+func UpgradeCluster(
+	client clientv1.Client,
+	dc dynamic.Interface,
+	clusterName string,
+	curator *clustercuratorv1.ClusterCurator) error {
+	klog.V(0).Info("* Initiate Upgrade")
+	klog.V(2).Info("Looking up managedclusterinfo " + clusterName)
+
+	desiredUpdate := curator.Spec.Upgrade.DesiredUpdate
+
+	if err := validateUpgradeVersion(client, clusterName, curator, desiredUpdate); err != nil {
+		return err
+	}
+
+	// Only support upgrading HC AND NPs to same versions for now
+	image := "quay.io/openshift-release-dev/ocp-release:" + desiredUpdate + "-multi"
+
+	// Patch HostedCluster with new image
+	err := patchUpgradeVersion(dc, clusterName, curator.Namespace, utils.HCGVR, image)
+	utils.CheckError(err)
+
+	// Need to account for 0 or multiple NodePools
+	nodePools, err := dc.Resource(utils.NPGVR).Namespace(curator.Namespace).List(context.TODO(), v1.ListOptions{})
+	utils.CheckError(err)
+
+	for _, np := range nodePools.Items {
+		spec := np.Object["spec"].(map[string]interface{})
+		npClusterName := spec["clusterName"].(string)
+		if npClusterName == clusterName {
+			npName := np.Object["metadata"].(map[string]interface{})["name"].(string)
+			err = patchUpgradeVersion(dc, npName, curator.Namespace, utils.NPGVR, image)
+			utils.CheckError(err)
+		}
+	}
+
+	return nil
+}
+
+func MonitorUpgradeStatus(
+	dc dynamic.Interface,
+	client clientv1.Client,
+	clusterName string,
+	curator *clustercuratorv1.ClusterCurator) error {
+	upgradeAttempts := utils.GetRetryTimes(curator.Spec.Upgrade.MonitorTimeout, 120, utils.PauseSixtySeconds)
+	klog.V(0).Info("Monitoring up to " + strconv.Itoa(upgradeAttempts) + " attempts for Hypershift Upgrade job")
+	elapsedTime := 0
+
+	// Refresh the hostedCluster resource
+	hostedCluster, err := dc.Resource(utils.HCGVR).Namespace(curator.Namespace).Get(
+		context.TODO(), clusterName, v1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < upgradeAttempts; i++ {
+		if isHostedReady(hostedCluster, true) {
+			klog.V(2).Info("Upgrade succeeded ✓")
+			utils.CheckError(utils.RecordCurrentStatusCondition(
+				client,
+				clusterName,
+				curator.Namespace,
+				"hypershift-upgrade-job",
+				v1.ConditionTrue,
+				"upgrade-job"))
+
+			return nil
+		} else if !isHostedReady(hostedCluster, true) {
+			for !isHostedReady(hostedCluster, true) {
+				if elapsedTime%6 == 0 {
+					klog.V(0).Info("Upgrade Job:  - " + strconv.Itoa(elapsedTime/6) + "min")
+				}
+				time.Sleep(utils.PauseTenSeconds) // 10s
+				elapsedTime++
+
+				// Reset hostedCluster, so we make sure we're getting clean data (not cached)
+				hostedCluster, err = dc.Resource(utils.HCGVR).Namespace(curator.Namespace).Get(
+					context.TODO(), clusterName, v1.GetOptions{})
+
+				if err != nil && k8serrors.IsNotFound(err) {
+					break
+				}
+
+				utils.CheckError(err)
+			}
+		}
+	}
+
+	if hostedCluster != nil {
+		klog.Warning(hostedCluster.Object["status"].(map[string]interface{})["conditions"])
```

Comment:
Minor nit. Not sure if this will print out nicely so maybe is better if isHostedReady return a string that states why its not ready so you can use it here. Up to you if you think it's worth it. (Probably not worth it).